[Bug]: datanode panics when creating a volume discount program

### Problem encountered

Integrating volume discount programs into market-sim fuzz tests, datanode panics when proposing a volume discount program.
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x10211a8c8]

goroutine 206 [running]:
code.vegaprotocol.io/vega/datanode/service.(*VolumeDiscountPrograms).AddVolumeDiscountProgram(0x140036d7828?, {0x103c81ed8?, 0x14001823ec0?}, 0x1009c8ce4?)
	<autogenerated>:1 +0x28
code.vegaprotocol.io/vega/datanode/sqlsubscribers.(*VolumeDiscountProgram).consumeVolumeDiscountProgramStartedEvent(0x14000e47a70, {0x103c81ed8, 0x14001823ec0}, {0x10e9f1b38, 0x140047fe3b0})
	/Users/caed/vega-market-sim/extern/vega/datanode/sqlsubscribers/volume_discount_programs.go:69 +0x1cc
code.vegaprotocol.io/vega/datanode/sqlsubscribers.(*VolumeDiscountProgram).Push(0x1036a6160?, {0x103c81ed8, 0x14001823ec0}, {0x103c98b30?, 0x140047fe3b0})
	/Users/caed/vega-market-sim/extern/vega/datanode/sqlsubscribers/volume_discount_programs.go:57 +0x12c
code.vegaprotocol.io/vega/datanode/broker.(*SQLStoreBroker).push(0x1400014d680, {0x103c81ed8, 0x14001823ec0}, {0x103c81d50?, 0x14000e47a70}, {0x103c98b30?, 0x140047fe3b0?})
	/Users/caed/vega-market-sim/extern/vega/datanode/broker/sqlstore_broker.go:345 +0x1e4
code.vegaprotocol.io/vega/datanode/broker.(*SQLStoreBroker).handleEvent(0x1400014d680, {0x103c81ed8, 0x14001823ec0}, {0x103c98b30?, 0x140047fe3b0?})
	/Users/caed/vega-market-sim/extern/vega/datanode/broker/sqlstore_broker.go:333 +0x190
code.vegaprotocol.io/vega/datanode/broker.(*SQLStoreBroker).processBlock(0x1400014d680, {0x103c81e30, 0x140005b1810}, {0x103c81ed8, 0x1400162cdb0}, 0x14004755780, 0x1400061bbc0, 0x1400061bc20)
	/Users/caed/vega-market-sim/extern/vega/datanode/broker/sqlstore_broker.go:291 +0x9f4
code.vegaprotocol.io/vega/datanode/broker.(*SQLStoreBroker).Receive(0x1400014d680, {0x103c81e30, 0x140005b1810})
	/Users/caed/vega-market-sim/extern/vega/datanode/broker/sqlstore_broker.go:131 +0x120
code.vegaprotocol.io/vega/cmd/data-node/commands/start.(*NodeCommand).runNode.func6()
	/Users/caed/vega-market-sim/extern/vega/cmd/data-node/commands/start/node.go:152 +0x64
golang.org/x/sync/errgroup.(*Group).Go.func1()
	/Users/caed/go/pkg/mod/golang.org/x/sync@v0.1.0/errgroup/errgroup.go:75 +0x5c
created by golang.org/x/sync/errgroup.(*Group).Go
	/Users/caed/go/pkg/mod/golang.org/x/sync@v0.1.0/errgroup/errgroup.go:72 +0xa0
```

vega-market-sim dump (logs, config, replay, etc.)
[vega-sim-pyvt252q.zip](https://github.com/vegaprotocol/vega/files/12767785/vega-sim-pyvt252q.zip)


### Observed behaviour

-

### Expected behaviour

-

### Steps to reproduce

```bash
1. cd vega-market-sim
1. git checkout 
1. poetry shell
1. pytest -s tests/integration/test_fuzz.py
```


### Software version

6608e3a7862d6073900786d7fbb5531aea85bfac

### Failing test

tests/integration/test_fuzz.py

### Jenkins run

_No response_

### Configuration used

```shell
{
  "genesis_time": "2023-05-09T10:00:00.000000000Z",
  "chain_id": "CUSTOM",
  "initial_height": "0",
  "consensus_params": {
    "block": {
      "max_bytes": "22020096",
      "max_gas": "-1",
      "time_iota_ms": "1"
    },
    "evidence": {
      "max_age_num_blocks": "100000",
      "max_age_duration": "172800000000000",
      "max_bytes": "1048576"
    },
    "validator": {
      "pub_key_types": ["ed25519"]
    },
    "version": {}
  },
  "validators": [
    {
      "address": "7DFD3744A773C166B36F170E4F5EF3E998703F4E",
      "pub_key": {
        "type": "tendermint/PubKeyEd25519",
        "value": "2lPO6I0p5uCBZzeoA2+ECJ2+57LtLVMjXpxH5za9TvE="
      },
      "power": "10",
      "name": ""
    }
  ],
  "app_hash": "",
  "app_state": {
    "assets": {
      "VOTE": {
        "name": "VOTE",
        "symbol": "VOTE",
        "decimals": 5,
        "min_lp_stake": "1",
        "source": {
          "builtin_asset": {
            "max_faucet_amount_mint": "100000000000"
          }
        }
      }
    },
    "validators": {
      "6NemqXSfN4zfZvAQeIANflEFqQ98izX9C0KLI2dITSY=": {
        "id": "c2b0ab69fb1bd19862c04df401920ea62d3e0bce41a50148240f42017824d17a",
        "vega_pub_key": "dfcf0fa8d67cfa11b3a511e320f3c47807a5e0735cc06dd9c2bc00f7ccc67117",
        "vega_pub_key_index": 1,
        "ethereum_address": "0x9e5BEEC6E56B28cCbd02864840B0f1e0125e42Ce",
        "tm_pub_key": "6NemqXSfN4zfZvAQeIANflEFqQ98izX9C0KLI2dITSY=",
        "info_url": "",
        "country": "",
        "name": "",
        "avatar_url": ""
      }
    },
    "network": {
      "replay_attack_threshold": 150
    },
    "network_parameters": {
      "blockchains.ethereumConfig": "{\"network_id\": \"3\", \"chain_id\": \"3\", \"collateral_bridge_contract\": { \"address\": \"0xa6F1E140daC13002Dfd9789D6dBA59117c717D7a\" }, \"confirmations\": 50, \"staking_bridge_contract\": { \"address\": \"0xfce2CC92203A266a9C8e67461ae5067c78f67235\", \"deployment_block_height\": 11001702}, \"multisig_control_contract\": {\"address\": \"0xCF6d41235911184fe6F35D47207813bFF3B91601\", \"deployment_block_height\": 12710009 } }",
      "governance.proposal.asset.maxClose": "8760h0m0s",
      "governance.proposal.asset.maxEnact": "8760h0m0s",
      "governance.proposal.asset.minClose": "1s",
      "governance.proposal.asset.minEnact": "1s",
      "governance.proposal.asset.minProposerBalance": "1",
      "governance.proposal.asset.minVoterBalance": "1",
      "governance.proposal.asset.requiredMajority": "0.66",
      "governance.proposal.asset.requiredParticipation": "0.00001",
      "governance.proposal.freeform.maxClose": "8760h0m0s",
      "governance.proposal.freeform.minClose": "1s",
      "governance.proposal.freeform.minProposerBalance": "1",
      "governance.proposal.freeform.minVoterBalance": "1",
      "governance.proposal.freeform.requiredMajority": "0.66",
      "governance.proposal.freeform.requiredParticipation": "0.00001",
      "governance.proposal.market.maxClose": "8760h0m0s",
      "governance.proposal.market.maxEnact": "8760h0m0s",
      "governance.proposal.market.minClose": "1s",
      "governance.proposal.market.minEnact": "2s",
      "governance.proposal.market.minProposerBalance": "1",
      "governance.proposal.market.minVoterBalance": "1",
      "governance.proposal.market.requiredMajority": "0.66",
      "governance.proposal.market.requiredParticipation": "0.00001",
      "governance.proposal.updateAsset.maxClose": "720h",
      "governance.proposal.updateAsset.maxEnact": "720h",
      "governance.proposal.updateAsset.minClose": "1s",
      "governance.proposal.updateAsset.minEnact": "1s",
      "governance.proposal.updateAsset.minProposerBalance": "1",
      "governance.proposal.updateAsset.minVoterBalance": "1",
      "governance.proposal.updateAsset.requiredMajority": "0.66",
      "governance.proposal.updateAsset.requiredParticipation": "0.09",
      "governance.proposal.updateMarket.maxClose": "8760h0m0s",
      "governance.proposal.updateMarket.maxEnact": "8760h0m0s",
      "governance.proposal.updateMarket.minClose": "1s",
      "governance.proposal.updateMarket.minEnact": "1s",
      "governance.proposal.updateMarket.minProposerBalance": "1",
      "governance.proposal.updateMarket.minProposerEquityLikeShare": "0",
      "governance.proposal.updateMarket.minVoterBalance": "1",
      "governance.proposal.updateMarket.requiredMajority": "0.66",
      "governance.proposal.updateMarket.requiredMajorityLP": "0.66",
      "governance.proposal.updateMarket.requiredParticipation": "0.00001",
      "governance.proposal.updateMarket.requiredParticipationLP": "0",
      "governance.proposal.updateNetParam.maxClose": "8760h0m0s",
      "governance.proposal.updateNetParam.maxEnact": "8760h0m0s",
      "governance.proposal.updateNetParam.minClose": "1s",
      "governance.proposal.updateNetParam.minEnact": "1s",
      "governance.proposal.updateNetParam.minProposerBalance": "1",
      "governance.proposal.updateNetParam.minVoterBalance": "1",
      "governance.proposal.updateNetParam.requiredMajority": "0.5",
      "governance.proposal.updateNetParam.requiredParticipation": "0.00001",
      "governance.proposal.referralProgram.maxClose": "8760h0m0s",
      "governance.proposal.referralProgram.maxEnact": "8760h0m0s",
      "governance.proposal.referralProgram.minClose": "1s",
      "governance.proposal.referralProgram.minEnact": "1s",
      "governance.proposal.referralProgram.minProposerBalance": "1",
      "governance.proposal.referralProgram.minVoterBalance": "1",
      "governance.proposal.referralProgram.requiredMajority": "0.5",
      "governance.proposal.referralProgram.requiredParticipation": "0.00001",
      "governance.proposal.VolumeDiscountProgram.maxClose": "8760h0m0s",
      "governance.proposal.VolumeDiscountProgram.maxEnact": "8760h0m0s",
      "governance.proposal.VolumeDiscountProgram.minClose": "1s",
      "governance.proposal.VolumeDiscountProgram.minEnact": "1s",
      "governance.proposal.VolumeDiscountProgram.minProposerBalance": "1",
      "governance.proposal.VolumeDiscountProgram.minVoterBalance": "1",
      "governance.proposal.VolumeDiscountProgram.requiredMajority": "0.5",
      "governance.proposal.VolumeDiscountProgram.requiredParticipation": "0.00001",
      "limits.markets.maxPeggedOrders": "1500",
      "market.auction.maximumDuration": "168h",
      "market.auction.minimumDuration": "1s",
      "market.fee.factors.infrastructureFee": "0.0005",
      "market.fee.factors.makerFee": "0.0002",
      "market.liquidity.bondPenaltyParameter": "0.1",
      "market.liquidity.earlyExitPenalty": "0.1",
      "market.liquidity.maximumLiquidityFeeFactorLevel": "0.03",
      "market.liquidity.minimum.probabilityOfTrading.lpOrders": "1e-6",
      "market.liquidity.probabilityOfTrading.tau.scaling": "10.0",
      "market.liquidity.providersFeeCalculationTimeStep": "1s",
      "market.liquidity.sla.nonPerformanceBondPenaltyMax": "0.5",
      "market.liquidity.sla.nonPerformanceBondPenaltySlope": "2",
      "market.liquidity.stakeToCcyVolume": "1.0",
      "market.liquidity.targetstake.triggering.ratio": "0.25",
      "market.liquidityProvision.minLpStakeQuantumMultiple": "5000",
      "market.liquidityProvision.shapes.maxSize": "100",
      "market.margin.scalingFactors": "{\"search_level\": 1.1, \"initial_margin\": 1.5, \"collateral_release\": 1.7}",
      "market.monitor.price.defaultParameters": "{\"triggers\": [{\"auction_extension\": 300, \"horizon\": 43200, \"probability\": \"0.9999999\"}] }",
      "market.stake.target.scalingFactor": "0.0001",
      "market.stake.target.timeWindow": "1h",
      "market.value.windowLength": "2h0m0s",
      "network.checkpoint.timeElapsedBetweenCheckpoints": "5m",
      "network.floatingPointUpdates.delay": "5m",
      "network.markPriceUpdateMaximumFrequency": "5s",
      "network.transaction.defaultgas": "1",
      "network.transactions.maxgasperblock": "3000000",
      "network.transactions.minBlockCapacity": "32",
      "network.validators.ersatz.multipleOfTendermintValidators": "0",
      "network.validators.ersatz.rewardFactor": "0.85",
      "network.validators.incumbentBonus": "0.05",
      "network.validators.minimumEthereumEventsForNewValidator": "3",
      "network.validators.multisig.numberOfSigners": "13",
      "network.validators.tendermint.number": "13",
      "reward.asset": "VOTE",
      "reward.staking.delegation.competitionLevel": "1.3",
      "reward.staking.delegation.delegatorShare": "0.8",
      "reward.staking.delegation.maxPayoutPerEpoch": "9890000000000000000000",
      "reward.staking.delegation.maxPayoutPerParticipant": "700000000000000000000",
      "reward.staking.delegation.minimumValidatorStake": "3000000000000000000000",
      "reward.staking.delegation.minValidators": "5",
      "reward.staking.delegation.optimalStakeMultiplier": "3.0",
      "reward.staking.delegation.payoutDelay": "0h",
      "reward.staking.delegation.payoutFraction": "1.0",
      "rewards.marketCreationQuantumMultiple": "50000000",
      "snapshot.interval.length": "300",
      "spam.pow.difficulty": "15",
      "spam.pow.hashFunction": "sha3_24_rounds",
      "spam.pow.increaseDifficulty": "0",
      "spam.pow.numberOfPastBlocks": "100",
      "spam.pow.numberOfTxPerBlock": "1000",
      "spam.protection.delegation.min.tokens": "100000000000000000",
      "spam.protection.max.batchSize": "30",
      "spam.protection.max.delegations": "360",
      "spam.protection.max.proposals": "300",
      "spam.protection.max.votes": "300",
      "spam.protection.maxUserTransfersPerEpoch": "20",
      "spam.protection.minimumWithdrawalQuantumMultiple": "10",
      "spam.protection.minMultisigUpdates": "100000000000000000000",
      "spam.protection.proposal.min.tokens": "2000000000000000000000",
      "spam.protection.voting.min.tokens": "1000000000000000000",
      "transfer.fee.factor": "0.001",
      "transfer.minTransferQuantumMultiple": "100",
      "validator.performance.scaling.factor": "0",
      "validators.delegation.minAmount": "1",
      "validators.epoch.length": "5m",
      "validators.vote.required": "0.67",
      "referralProgram.maxReferralTiers": "10",
      "referralProgram.maxReferralRewardFactor": "1",
      "referralProgram.maxReferralDiscountFactor": "1",
      "referralProgram.maxPartyNotionalVolumeByQuantumPerEpoch": "10000",
      "referralProgram.minStakedVegaTokens": "0",
      "referralProgram.maxReferralRewardProportion": "1"
    },
    "network_limits": {
      "propose_market_enabled": true,
      "propose_asset_enabled": true,
      "bootstrap_block_count": 0
    },
    "checkpoint": {
      "load_hash": ""
    }
  }
}
```


### Relevant log output

```shell
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x10211a8c8]

goroutine 206 [running]:
code.vegaprotocol.io/vega/datanode/service.(*VolumeDiscountPrograms).AddVolumeDiscountProgram(0x140036d7828?, {0x103c81ed8?, 0x14001823ec0?}, 0x1009c8ce4?)
	<autogenerated>:1 +0x28
code.vegaprotocol.io/vega/datanode/sqlsubscribers.(*VolumeDiscountProgram).consumeVolumeDiscountProgramStartedEvent(0x14000e47a70, {0x103c81ed8, 0x14001823ec0}, {0x10e9f1b38, 0x140047fe3b0})
	/Users/caed/vega-market-sim/extern/vega/datanode/sqlsubscribers/volume_discount_programs.go:69 +0x1cc
code.vegaprotocol.io/vega/datanode/sqlsubscribers.(*VolumeDiscountProgram).Push(0x1036a6160?, {0x103c81ed8, 0x14001823ec0}, {0x103c98b30?, 0x140047fe3b0})
	/Users/caed/vega-market-sim/extern/vega/datanode/sqlsubscribers/volume_discount_programs.go:57 +0x12c
code.vegaprotocol.io/vega/datanode/broker.(*SQLStoreBroker).push(0x1400014d680, {0x103c81ed8, 0x14001823ec0}, {0x103c81d50?, 0x14000e47a70}, {0x103c98b30?, 0x140047fe3b0?})
	/Users/caed/vega-market-sim/extern/vega/datanode/broker/sqlstore_broker.go:345 +0x1e4
code.vegaprotocol.io/vega/datanode/broker.(*SQLStoreBroker).handleEvent(0x1400014d680, {0x103c81ed8, 0x14001823ec0}, {0x103c98b30?, 0x140047fe3b0?})
	/Users/caed/vega-market-sim/extern/vega/datanode/broker/sqlstore_broker.go:333 +0x190
code.vegaprotocol.io/vega/datanode/broker.(*SQLStoreBroker).processBlock(0x1400014d680, {0x103c81e30, 0x140005b1810}, {0x103c81ed8, 0x1400162cdb0}, 0x14004755780, 0x1400061bbc0, 0x1400061bc20)
	/Users/caed/vega-market-sim/extern/vega/datanode/broker/sqlstore_broker.go:291 +0x9f4
code.vegaprotocol.io/vega/datanode/broker.(*SQLStoreBroker).Receive(0x1400014d680, {0x103c81e30, 0x140005b1810})
	/Users/caed/vega-market-sim/extern/vega/datanode/broker/sqlstore_broker.go:131 +0x120
code.vegaprotocol.io/vega/cmd/data-node/commands/start.(*NodeCommand).runNode.func6()
	/Users/caed/vega-market-sim/extern/vega/cmd/data-node/commands/start/node.go:152 +0x64
golang.org/x/sync/errgroup.(*Group).Go.func1()
	/Users/caed/go/pkg/mod/golang.org/x/sync@v0.1.0/errgroup/errgroup.go:75 +0x5c
created by golang.org/x/sync/errgroup.(*Group).Go
	/Users/caed/go/pkg/mod/golang.org/x/sync@v0.1.0/errgroup/errgroup.go:72 +0xa0
```

--- a/cmd/data-node/commands/start/sqlsubscribers.go
+++ b/cmd/data-node/commands/start/sqlsubscribers.go
@@ -57,9 +57,9 @@ type SQLSubscribers struct {
 	referralSetsStore           *sqlstore.ReferralSets
 	teamsStore                  *sqlstore.Teams
 	vestingStatsStore           *sqlstore.VestingStats
-	volumeDiscountStatsStore    *sqlstore.VolumeDiscountStats
 	referralFeeStatsStore       *sqlstore.ReferralFeeStats
 	fundingPaymentStore         *sqlstore.FundingPayments
+	volumeDiscountStatsStore    *sqlstore.VolumeDiscountStats
 	volumeDiscountProgramsStore *sqlstore.VolumeDiscountPrograms
 
 	// Services
@@ -106,9 +106,9 @@ type SQLSubscribers struct {
 	referralSetsService          *service.ReferralSets
 	teamsService                 *service.Teams
 	vestingStatsService          *service.VestingStats
-	volumeDiscountStatsService   *service.VolumeDiscountStats
 	referralFeeStatsService      *service.ReferralFeeStats
 	fundingPaymentService        *service.FundingPayment
+	volumeDiscountStatsService   *service.VolumeDiscountStats
 	volumeDiscountProgramService *service.VolumeDiscountPrograms
 
 	// Subscribers
@@ -153,9 +153,9 @@ type SQLSubscribers struct {
 	referralSetsSub          *sqlsubscribers.ReferralSets
 	teamsSub                 *sqlsubscribers.Teams
 	vestingStatsSub          *sqlsubscribers.VestingStatsUpdated
-	volumeDiscountStatsSub   *sqlsubscribers.VolumeDiscountStatsUpdated
 	referralFeeStatsSub      *sqlsubscribers.ReferralFeeStats
 	fundingPaymentSub        *sqlsubscribers.FundingPaymentSubscriber
+	volumeDiscountStatsSub   *sqlsubscribers.VolumeDiscountStatsUpdated
 	volumeDiscountProgramSub *sqlsubscribers.VolumeDiscountProgram
 }
 
@@ -204,9 +204,9 @@ func (s *SQLSubscribers) GetSQLSubscribers() []broker.SQLBrokerSubscriber {
 		s.referralSetsSub,
 		s.teamsSub,
 		s.vestingStatsSub,
-		s.volumeDiscountStatsSub,
 		s.referralFeeStatsSub,
 		s.fundingPaymentSub,
+		s.volumeDiscountStatsSub,
 		s.volumeDiscountProgramSub,
 	}
 }
@@ -259,9 +259,9 @@ func (s *SQLSubscribers) CreateAllStores(ctx context.Context, Log *logging.Logge
 	s.referralSetsStore = sqlstore.NewReferralSets(transactionalConnectionSource)
 	s.teamsStore = sqlstore.NewTeams(transactionalConnectionSource)
 	s.vestingStatsStore = sqlstore.NewVestingStats(transactionalConnectionSource)
-	s.volumeDiscountStatsStore = sqlstore.NewVolumeDiscountStats(transactionalConnectionSource)
 	s.referralFeeStatsStore = sqlstore.NewReferralFeeStats(transactionalConnectionSource)
 	s.fundingPaymentStore = sqlstore.NewFundingPayments(transactionalConnectionSource)
+	s.volumeDiscountStatsStore = sqlstore.NewVolumeDiscountStats(transactionalConnectionSource)
 	s.volumeDiscountProgramsStore = sqlstore.NewVolumeDiscountPrograms(transactionalConnectionSource)
 }
 
@@ -309,9 +309,10 @@ func (s *SQLSubscribers) SetupServices(ctx context.Context, log *logging.Logger,
 	s.referralSetsService = service.NewReferralSets(s.referralSetsStore)
 	s.teamsService = service.NewTeams(s.teamsStore)
 	s.vestingStatsService = service.NewVestingStats(s.vestingStatsStore)
-	s.volumeDiscountStatsService = service.NewVolumeDiscountStats(s.volumeDiscountStatsStore)
 	s.referralFeeStatsService = service.NewReferralFeeStats(s.referralFeeStatsStore)
 	s.fundingPaymentService = service.NewFundingPayment(s.fundingPaymentStore)
+	s.volumeDiscountStatsService = service.NewVolumeDiscountStats(s.volumeDiscountStatsStore)
+	s.volumeDiscountProgramService = service.NewVolumeDiscountPrograms(s.volumeDiscountProgramsStore)
 
 	toInit := []interface{ Initialise(context.Context) error }{
 		s.marketDepthService,
@@ -370,8 +371,8 @@ func (s *SQLSubscribers) SetupSQLSubscribers() {
 	s.referralSetsSub = sqlsubscribers.NewReferralSets(s.referralSetsService)
 	s.teamsSub = sqlsubscribers.NewTeams(s.teamsService)
 	s.vestingStatsSub = sqlsubscribers.NewVestingStatsUpdated(s.vestingStatsService)
-	s.volumeDiscountStatsSub = sqlsubscribers.NewVolumeDiscountStatsUpdated(s.volumeDiscountStatsService)
 	s.referralFeeStatsSub = sqlsubscribers.NewReferralFeeStats(s.referralFeeStatsService)
 	s.fundingPaymentSub = sqlsubscribers.NewFundingPaymentsSubscriber(s.fundingPaymentStore)
+	s.volumeDiscountStatsSub = sqlsubscribers.NewVolumeDiscountStatsUpdated(s.volumeDiscountStatsService)
 	s.volumeDiscountProgramSub = sqlsubscribers.NewVolumeDiscountProgram(s.volumeDiscountProgramService)
 }

